### PR TITLE
ANSI libsa functions

### PR DESCRIPTION
**Pull request by AtariDreams**
A lot of the libsa functions were written back when ANSI C wasn’t a thing. So, I replaced their old, nonstandard parts with their modern-day equivalents.

No functional change intended.

### CODE REVIEW

File: stand/libsa/in_cksum.c [hunk 1]
```diff
@@ -79,7 +77,7 @@ in_cksum(p, len)
 #endif
 			}
 		}
-		if ((oddbyte = len & 1) != 0)
+		if ((oddbyte = len & 1))
```
Comment (dsdqmhsx) [inline stand/libsa/in_cksum.c:80]: unrelated and wrong?

File: stand/libsa/ether.c [hunk 2]
```diff
@@ -133,15 +133,15 @@ static char digits[] = "0123456789abcdef";
 char *
 ether_sprintf(u_char *ap)
 {
-	int i;
+	size_t i;
 	static char etherbuf[18];
 	char *cp = etherbuf;
 
-	for (i = 0; i < 6; i++) {
+	for (i = 0; i < 6; ++i) {
```
Comment (dsdqmhsx) [inline stand/libsa/ether.c:140]: unrelated

**Review by dsdqmhsx — COMMENTED** [review]
while this does ansi'fy the code, it also seems to contain a lot of unrelated and questionable changes

**Comment by gAlfonso-bit** [comment]
> while this does ansi'fy the code, it also seems to contain a lot of unrelated and questionable changes

Fixed

**Comment by AtariDreams** [comment]
Any updates? It has been 30 days @bsdimp @dsdqmhsx

**Comment by bsdimp** [comment]
I've been on vacation and/or tied up lately.  I'll take a look at this and other pull requests later in the week.

**Comment by AtariDreams** [comment]
> I've been on vacation and/or tied up lately. I'll take a look at this and other pull requests later in the week.

Alright

**Comment by bsdimp** [comment]
Let's start with this one...

File: stand/libsa/in_cksum.c [hunk 3]
```diff
@@ -50,20 +50,18 @@ __FBSDID("$FreeBSD$");
...
+	u_char *cp = (u_char *)p;
 
 	/* we assume < 2^16 bytes being summed */
 	while (len > 0) {
 		if (oddbyte) {
 			sum += v + *cp++;
 			len--;
 		}
-		if (((long)cp & 1) == 0) {
+		if (((intptr_t)cp & 1) == 0) {
```
Comment (bsdimp) [inline stand/libsa/in_cksum.c:64]: This change is good (arguably it should be uintptr_t, but either are good). It should be a separate commit.

File: stand/libsa/in_cksum.c [hunk 4]
```diff
@@ -90,5 +88,5 @@ in_cksum(p, len)
 		sum += v;
 	sum = (sum >> 16) + (sum & 0xffff); /* add in accumulated carries */
 	sum += sum >> 16;		/* add potential last carry */
-	return (0xffff & ~sum);
+	return (~sum & 0xffff);
```
Comment (bsdimp) [inline stand/libsa/in_cksum.c:91]: This commit is correct, but it's also unnecessary. Unless there's a really good reason for it, I'd drop it.

File: stand/libsa/stand.h [hunk 5]
```diff
@@ -352,7 +352,7 @@ extern ev_sethook_t	env_noset;		/* refuse set operation */
 extern ev_unsethook_t	env_nounset;		/* refuse unset operation */
 
 /* stdlib.h routines */
-extern int		abs(int a);
+extern int		abs(int a) __pure2;
```
Comment (bsdimp) [inline stand/libsa/stand.h:355]: Also a separate commit. It may be correct, but it's unrelated to the K&R function definitions -> ANSI prototypes that's the main thrust of this change.

File: stand/libsa/strcasecmp.c [hunk 6]
```diff
@@ -39,34 +39,39 @@ static char sccsid[] = "@(#)strcasecmp.c	8.1 (Berkeley) 6/4/93";
...
+	u_char u1, u2;
+
+	while ((u1 = tolower(*us1)) == (u2 = tolower(*us2))) {
+		if (u1 == '\0') {
 			return (0);
-	return (tolower(*us1) - tolower(*--us2));
+		}
+		us1++, us2++;
+	}
+	return (u1 - u2);
```
Comment (bsdimp) [inline stand/libsa/strcasecmp.c:55]: All of these changes should likely be in a separate pull request. It's not a simple cleanup, but completely rewrites this function. I'm honestly not sure that benefit from having the extra intermediaries, but it seems different enough to move to it's own pull request. The other cleanup could go together with the ANSI prototypes stuff, but this goes beyond a tweak.

File: stand/libsa/strcasecmp.c [hunk 7]
```diff
@@ -39,34 +39,39 @@ static char sccsid[] = "@(#)strcasecmp.c	8.1 (Berkeley) 6/4/93";
...
+			if ((u1 = tolower(*us1)) != (u2 = tolower(*us2))) {
+				return (u1 - u2);
+			}
+			if (u1 == '\0') {
 				break;
+			}
+			us1++, us2++;
 		} while (--n != 0);
 	}
 	return (0);
```
Comment (bsdimp) [inline stand/libsa/strcasecmp.c:77]: Same comment here as for strcasecmp.

File: stand/libsa/strdup.c [hunk 8]
```diff
@@ -39,17 +39,15 @@ static char sccsid[] = "@(#)strdup.c	8.1 (Berkeley) 6/4/93";
...
+	size_t len;
 
 	if (str != NULL) {
-	    len = strlen(str) + 1;
-	    if ((copy = malloc(len)) == NULL)
-		return (NULL);
-	    memcpy(copy, str, len);
+		len = strlen(str) + 1;
+		if ((copy = (char *)malloc(len)) != NULL)
+			memcpy(copy, str, len);
```
Comment (bsdimp) [inline stand/libsa/strdup.c:50]: This should be it's own commit. It's basically just a whitespace change.

File: stand/libsa/strdup.c [hunk 9]
```diff
@@ -39,17 +39,15 @@ static char sccsid[] = "@(#)strdup.c	8.1 (Berkeley) 6/4/93";
 #include <string.h>
 
 char *
-strdup(str)
-	const char *str;
+strdup(const char *str)
 {
-	size_t len;
 	char *copy = NULL;
+	size_t len;
```
Comment (bsdimp) [inline stand/libsa/strdup.c:45]: There's no reason to move this.

File: stand/libsa/strdup.c [hunk 10]
```diff
@@ -39,17 +39,15 @@ static char sccsid[] = "@(#)strdup.c	8.1 (Berkeley) 6/4/93";
...
 
 	if (str != NULL) {
-	    len = strlen(str) + 1;
-	    if ((copy = malloc(len)) == NULL)
-		return (NULL);
-	    memcpy(copy, str, len);
+		len = strlen(str) + 1;
+		if ((copy = (char *)malloc(len)) != NULL)
```
Comment (bsdimp) [inline stand/libsa/strdup.c:49]: You don't need this cast.

File: stand/libsa/ufs.c [hunk 11]
```diff
@@ -207,10 +205,7 @@ read_inode(inumber, f)
  * contains that block.
  */
 static int
-block_map(f, file_block, disk_block_p)
-	struct open_file *f;
-	ufs2_daddr_t file_block;
-	ufs2_daddr_t *disk_block_p;	/* out */
+block_map(struct open_file *f, ufs2_daddr_t file_block, ufs2_daddr_t *disk_block_p)	/* out */
```
Comment (bsdimp) [inline stand/libsa/ufs.c:208]: This line may now be > 80 characters. Also the comment at the end doesn't make sense. It was bound to the disk_block_p function argument (it's output by block_map).

File: stand/libsa/ufs.c [hunk 12]
```diff
@@ -312,10 +307,7 @@ block_map(f, file_block, disk_block_p)
  * Write a portion of a file from an internal buffer.
  */
 static int
-buf_write_file(f, buf_p, size_p)
-	struct open_file *f;
-	const char *buf_p;
-	size_t *size_p;		/* out */
+buf_write_file(struct open_file *f, const char *buf_p, size_t *size_p)	/* out */
```
Comment (bsdimp) [inline stand/libsa/ufs.c:310]: As as my comment for block_map above

File: stand/libsa/ufs.c [hunk 13]
```diff
@@ -390,10 +382,7 @@ buf_write_file(f, buf_p, size_p)
  * the location in the buffer and the amount in the buffer.
  */
 static int
-buf_read_file(f, buf_p, size_p)
-	struct open_file *f;
-	char **buf_p;		/* out */
-	size_t *size_p;		/* out */
+buf_read_file(struct open_file *f, char **buf_p, size_t *size_p)		/* out */
```
Comment (bsdimp) [inline stand/libsa/ufs.c:385]: Same as my comment for block_map above (though maybe this one isn't > 80 characters.

File: stand/libsa/ufs.c [hunk 14]
```diff
@@ -408,7 +397,7 @@ buf_read_file(f, buf_p, size_p)
 	block_size = sblksize(fs, DIP(fp, di_size), file_block);
 
 	if (file_block != fp->f_buf_blkno) {
-		if (fp->f_buf == (char *)0)
+		if (fp->f_buf == NULL)
```
Comment (bsdimp) [inline stand/libsa/ufs.c:400]: This should be a separate commit.

File: stand/libsa/ufs.c [hunk 15]
```diff
@@ -452,10 +441,7 @@ buf_read_file(f, buf_p, size_p)
  * i_number.
  */
 static int
-search_directory(name, f, inumber_p)
-	char *name;
-	struct open_file *f;
-	ino_t *inumber_p;		/* out */
+search_directory(char *name, struct open_file *f, ino_t *inumber_p)		/* out */
```
Comment (bsdimp) [inline stand/libsa/ufs.c:444]: Same comment as block_map above.

File: stand/libsa/ufs.c [hunk 16]
```diff
@@ -741,11 +724,7 @@ ufs_close(f)
  * Cross block boundaries when necessary.
  */
 static int
-ufs_read(f, start, size, resid)
-	struct open_file *f;
-	void *start;
-	size_t size;
-	size_t *resid;	/* out */
+ufs_read(struct open_file *f, void *start, size_t size, size_t *resid)	/* out */
```
Comment (bsdimp) [inline stand/libsa/ufs.c:727]: Same comment as block_map above.

File: stand/libsa/ufs.c [hunk 17]
```diff
@@ -880,16 +854,17 @@ ufs_readdir(struct open_file *f, struct dirent *d)
...
+	do {
+		if (fp->f_seekp >= DIP(fp, di_size))
+			return (ENOENT);
+		error = buf_read_file(f, &buf, &buf_size);
+		if (error)
+			return (error);
+		dp = (struct direct *)buf;
+		fp->f_seekp += dp->d_reclen;
+	} while (dp->d_ino == (ino_t)0);
+
```
Comment (bsdimp) [inline stand/libsa/ufs.c:867]: This should be a separate commit.

**Review by bsdimp — COMMENTED** [review]
I've added a lot of comments, mostly about splitting things up.  I can help you on discord with the mechanics of this if you need help with driving 'git' on the command line, or if you have other questions that aren't better hashed out here.

This submission on the whole looks mostly good. Separate commits will help the project in two ways. First, these cleanup commits have a small rate of bug introduction. Separate commits will help us bisect to find the troublesome one. Second, having smaller chunks means they are easier to merge into stable branches.

**Comment by AtariDreams** [comment]
Fixed!

File: stand/libsa/strdup.c [hunk 18]
```diff
@@ -39,17 +39,14 @@ static char sccsid[] = "@(#)strdup.c	8.1 (Berkeley) 6/4/93";
...
 	char *copy = NULL;
 
 	if (str != NULL) {
-	    len = strlen(str) + 1;
-	    if ((copy = malloc(len)) == NULL)
-		return (NULL);
-	    memcpy(copy, str, len);
+		const size_t len = strlen(str) + 1;
+		if ((copy = malloc(len)) != NULL)
+			memcpy(copy, str, len);
```
Comment (bsdimp) [inline stand/libsa/strdup.c:49]: This bit shouldn't be here. It's unreleated.

**Comment by AtariDreams** [comment]
Fixed!

File: stand/libsa/strcasecmp.c [hunk 19]
```diff
@@ -39,8 +39,7 @@ static char sccsid[] = "@(#)strcasecmp.c	8.1 (Berkeley) 6/4/93";
 #endif /* LIBC_SCCS and not lint */
 
 int
-strcasecmp(s1, s2)
-	const char *s1, *s2;
+strcasecmp(const char *s1, const char *s2, size_t n)
```
Comment (bsdimp) [inline stand/libsa/strcasecmp.c:42]: this adds an extra size_t n.
Fixed it in the pull

**Comment by bsdimp** [comment]
Corrected version pulled as 1f629966d683ca2e79adbc6c39e7ec5f1a87832f
